src/calling/haplotypes.rs

```diff
@@ -177,49 +184,93 @@ impl KallistoEstimates {
 pub(crate) struct Variant(#[deref] String);
 
 #[derive(Derefable, Debug, Clone, PartialEq, Eq)]
-pub(crate) struct HaplotypeVariants(#[deref] HashMap<Variant, Vec<Haplotype>>);
+pub(crate) struct HaplotypeVariants(#[deref] HashMap<Variant, BitVec>);
 
 impl HaplotypeVariants {
-    pub(crate) fn new(haplotype_variants: &mut bcf::Reader) -> Result<Vec<HaplotypeVariants>> {
-        let mut vec_haplotype_variants = Vec::new();
+    pub(crate) fn new(haplotype_variants: &mut bcf::Reader) -> Result<Self> {
+        let mut candidate_variants = HashMap::new();
         for record_result in haplotype_variants.records() {
-            let record = record_result.expect("Failed to read record");
+            let record = record_result?;
 
             //store the variant
             let mut variant = String::new();
             variant.push_str(&format!("{}{}", record.contig(), "_"));
             let pos = record.pos().to_string();
             variant.push_str(&format!("{}{}", pos, "_"));
             for allele in record.alleles() {
-                for c in allele {
-                    let base = char::from(*c).to_string();
-                    variant.push_str(&format!("{}{}", base, "_"));
+                if allele.len() == 1 {
+                    for c in allele {
+                        let base = char::from(*c).to_string();
+                        variant.push_str(&format!("{}{}", base, "_"));
+                    }
+                } else {
+                    //the case for structural variants
+                    for c in allele {
+                        let base = char::from(*c).to_string();
+                        variant.push_str(&base);
+                    }
+                    variant.push_str("_")
                 }
             }
```

Comment:
I don't think we need to store variants like this. Granted you need a correspondence between the BCF with the varlociraptor calls and the BCF with the variant to haplotype assignment. But instead of building a key for this from scratch like you do here, it is better to use `record.id()`. For this to work, you just need to ensure that your variant to haplotype BCF contains a unique ID for each row (in the ID column of the BCF). This can be done during generation of that file (e.g. just increment a number).